tcms/core/tests/test_ajax.py

```diff
@@ -91,3 +98,92 @@ def test_update_default_tester_non_existing_user(self):
         self.assertEqual(result['response'], 'Default tester not found!')
 
         self._assert_default_tester_is(None)
+
+
+class Test_Tag_Add(test.TestCase):
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.url = reverse('ajax-tags')
+        cls.test_tag = TestTagFactory()
+        cls.test_plan = TestPlanFactory()
+        cls.test_case = TestCaseFactory()
+        cls.test_run = TestRunFactory()
+
+    def test_add_tag_to_test_plan(self):
+        response = self.client.get(self.url, {
+            'tags': self.test_tag,
+            'plan': self.test_plan.plan_id,
+            'a': 'add'
+        })
+
+        self.assertEqual(response.status_code, http.client.OK)
+        self.assertEqual(self.test_plan.tag.count(), 1)
+        self.assertTrue(self.test_tag in self.test_plan.tag.all())
+
+    def test_add_tag_to_test_case(self):
+        response = self.client.get(self.url, {
+            'tags': self.test_tag,
+            'case': self.test_case.case_id,
+            'a': 'add'
+        })
+
+        self.assertEqual(response.status_code, http.client.OK)
+        self.assertEqual(self.test_case.tag.count(), 1)
+        self.assertTrue(self.test_tag in self.test_case.tag.all())
+
+    def test_add_tag_to_test_run(self):
+        response = self.client.get(self.url, {
+            'tags': self.test_tag,
+            'run': self.test_run.run_id,
+            'a': 'add'
+        })
+
+        self.assertEqual(response.status_code, http.client.OK)
+        self.assertEqual(self.test_run.tag.count(), 1)
+        self.assertTrue(self.test_tag in self.test_run.tag.all())
+
+
+class Test_Tag_Remove(test.TestCase):
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.url = reverse('ajax-tags')
+        cls.test_tag = TestTagFactory()
+        cls.test_plan = TestPlanFactory()
+        cls.test_case = TestCaseFactory()
+        cls.test_run = TestRunFactory()
+
+        TestPlanTag(tag=cls.test_tag, plan=cls.test_plan).save()
+        TestCaseTag(tag=cls.test_tag, case=cls.test_case).save()
+        TestRunTag(tag=cls.test_tag, run=cls.test_run).save()
```

Comment:
These are intermediate model classes (for many to many or many to one relationships). You'd better use the respective `.add_tag` methods for TestPlan, TestCase and TestRun classes.